### PR TITLE
Contacts + Sessions + Account Info Dump

### PR DESCRIPTION
**Pull request by patrizok**

### CODE REVIEW

File: saver.go [hunk 1]
```diff
@@ -134,6 +146,17 @@ func (s JSONFilesHistorySaver) openForAppend(fpath string) (*os.File, error) {
 	return file, nil
 }
 
+func (s JSONFilesHistorySaver) openForCreate(fpath string) (*os.File, error) {
```
Comment (3bl3gamer) [inline saver.go:149]: Weird name. Maybe `openAndTruncate`?

File: saver.go [hunk 2]
```diff
@@ -317,6 +340,63 @@ func (s JSONFilesHistorySaver) SaveRelatedChats(chats []mtproto.TL) error {
 	return nil
 }
 
+func (s JSONFilesHistorySaver) SaveContacts(contacts []mtproto.TL) error {
+
+	var encoder *json.Encoder
+
+	if encoder == nil {
```
Comment (3bl3gamer) [inline saver.go:347]: It's always `nil`, so the check is unnecessary. It could be just `encoder := json.NewEncoder(file)`.

File: saver.go [hunk 3]
```diff
@@ -317,6 +340,63 @@ func (s JSONFilesHistorySaver) SaveRelatedChats(chats []mtproto.TL) error {
...
+	}
+
+	return nil
+}
+
+func (s JSONFilesHistorySaver) SaveAuths(auths []mtproto.TL) error {
+
+	var encoder *json.Encoder
+
+	if encoder == nil {
```
Comment (3bl3gamer) [inline saver.go:366]: Same as `SaveContacts`

File: saver.go [hunk 4]
```diff
@@ -317,6 +340,63 @@ func (s JSONFilesHistorySaver) SaveRelatedChats(chats []mtproto.TL) error {
...
+	}
+
+	return nil
+}
+
+func (s JSONFilesHistorySaver) SaveAccount(me mtproto.TL_user) error {
+
+	var encoder *json.Encoder
+
+	if encoder == nil {
```
Comment (3bl3gamer) [inline saver.go:385]: Same as `SaveContacts`

File: tg.go [hunk 5]
```diff
@@ -101,6 +101,10 @@ func tgConnect(config *Config, logHandler *LogHandler) (*tgclient.TGClient, erro
 		return nil, merry.Wrap(mtproto.WrongRespError(res))
 	}
 	me := users[0].(mtproto.TL_user)
+
+	saver := &JSONFilesHistorySaver{Dirpath: config.OutDirPath}
+	saver.SaveAccount(me)
```
Comment (3bl3gamer) [inline tg.go:106]: No saves should occur here. `tgConnect` should just connect. However, it can return `TL_user` as the second value (before error).

**Review by 3bl3gamer — CHANGES_REQUESTED** [review]
Hello! Thank you for the pull request!

But there are some problems.

## General

Account/contacts/auths should be configurable via `config.json`. Currently they will be saved on every run which may be undesirable: someone may have a lot of contacts but willing to update just one chat.

They must not update on `--list-chats`.

There should be an append option. Currently everything is appended: users, chats and (of course) messages. But account/contacts/auths will be completely overwritten. Maybe switch this via config? Like `{"contacts":"off"` / `{"contacts":"overwrite"` / `{"contacts":"append"`?

Since account/contacts/auths are saved as-is (exactly in format they come from TG), they should
a) have `"_"` field with type and
b) have `"_TL_LAYER"` field with API version (like messages: [README](https://github.com/3bl3gamer/tg-history-dumper#messages), [code example](https://github.com/3bl3gamer/tg-history-dumper/blob/master/saver.go#L334-L335)).

They should also have `"UpdatedAt"` field as users/chats do (at least in append mode).

P.S.: you are Russian, aren't you?

**Comment by patrizok** [comment]
Hi,
Thank you for your detailed suggestions.

• I update .gitignore file to prevent updating config and dump files.
• I resolve the problems that you point.
• I do not support updating session and contact right now. Because Telegram already return all data in one response. It can be disabled by config.json or command line parameters. UpdatedAt is not required also.
• Config.json supports enable and disable new features.
• Command line parameter enable disable works.
• Not update on --list-chats
• Document updated, but .zh document is not updated. You can update it later.

> Since account/contacts/auths are saved as-is (exactly in format they come from TG), they should
> a) have "_" field with type and
> b) have "_TL_LAYER" field with API version (like messages: README, code example).

• This is the only thing that I not implement. I don't know why users need the type and layer version.
Adding these information to dump files also increase code complexity, chats and users file also does not have this information.

I am not Russian by the way I'm from UK ;)

**Review by 3bl3gamer — CHANGES_REQUESTED** [review]
> .zh document is not updated

I dont't know Chinese either :) Just copy new parts (in English) to zh.md and maybe someone will translate them later.

> why users need the type and layer version
> chats and users file also does not have this information

That's because chats and users format is defined in [dumper's source](https://github.com/3bl3gamer/tg-history-dumper/blob/master/saver.go#L16-L23), have only limited amount of fields and most likely will not change.

[User](https://github.com/3bl3gamer/tgclient/blob/7cc3e61624657aae935e8d3c92ea8dd886a4da27/mtproto/tl_schema.go#L1991), [Contact](https://github.com/3bl3gamer/tgclient/blob/7cc3e61624657aae935e8d3c92ea8dd886a4da27/mtproto/tl_schema.go#L2701) and [Authorization](https://github.com/3bl3gamer/tgclient/blob/7cc3e61624657aae935e8d3c92ea8dd886a4da27/mtproto/tl_schema.go#L4071) (huge 32k-line pages, beware) on the other hand are defined by Telegram API and may change with a high probability in the next API version (especially User). Without `_TL_LAYER` some quirks with fields presence/absence checks may be required when processing saved jsons. There is also `TL_contact.Mutial` field which is either `TL_boolTrue` or `TL_boolFalse`, and without type annotation both will be just `{}` .

Type (`"_"`) here is mostly for consistency with other exported TG API structs (in messages).

> TL_contacts_contacts

There are just IDs with separate `TL_contacts_contacts.Users` array. It should also be saved via `SaveRelatedUsers` (in case a contact has empty message history, etc).

**Comment by patrizok** [comment]
Hi,

As you mentioned changes applied.

File: README.md [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# Telegram History Dumber
+# Telegram History Dumper
```
Comment (3bl3gamer) [inline README.md:1]: OMG o_O
thank you

**Review by 3bl3gamer — CHANGES_REQUESTED** [review]
Aaand what about this comment https://github.com/3bl3gamer/tg-history-dumper/pull/7#pullrequestreview-699377378 ?

**Comment by 3bl3gamer** [comment]
> rename tg.session.temp tg.session: The process cannot access

Fixed this one. You can check it by updating `github.com/3bl3gamer/tgclient` dependency to `v0.126.1`

File: README.md (hunk 6 above)
Comment (patrizok) [reply README.md:1]: :)

**Comment by patrizok** [comment]
I update depencencies

With new version of **github.com/3bl3gamer/tgclient** file IO problem fixed but there is an another problem with Windows. 
I can not understand problem with errorcode.

> [ERROR] :
> unexpected response: mtproto.TL_bad_msg_notification{BadMsgID:6982098918416156760, BadMsgSeqno:7, ErrorCode:33}
> 
> main.tgConnect
>         /home/kali/Desktop/tg-history-dumper/tg.go:101
> main.dump
>         /home/kali/Desktop/tg-history-dumper/main.go:242
> main.main
>         /home/kali/Desktop/tg-history-dumper/main.go:333
> runtime.main
>         /usr/lib/go-1.15/src/runtime/proc.go:204
> runtime.goexit
>         /usr/lib/go-1.15/src/runtime/asm_amd64.s:1374
--------------

_Id dump comes with the TL_contacts_contacts.Users -> Contacts array_ seems **not required** because even if I do not start chat contact already comes in Users array. (contact has empty message history, etc) hypotheses not work here. Users and Contacts already mingling with each other adding SaveRelatedUsers file will be puzzle if not really required.

> There are just IDs with separate TL_contacts_contacts.Users array. It should also be saved via SaveRelatedUsers (in case a contact has empty message history, etc).

-------------------

 Comment #7 ( "_TL_LAYER" and "_" ) done right now.
I am not good at dependency management with go, so can you please check the mod and sum files.

**Review by 3bl3gamer — CHANGES_REQUESTED** [review]
> unexpected response: mtproto.TL_bad_msg_notification

That's a bug with `tgclient`. It happens sometimes when client gets interrupted during connection. And gets fixed after 2-4 re-connections.

> TL_contacts_contacts.Users/Contacts 

Not sure about this one. So there are two options:

First is to save `TL_contacts_contacts.Users`. The result will a) be easier to process and b) contain some extra fields which are missing in `history_my/users` (last status, bot/scam flag, etc.)

Another one is to save `TL_contacts_contacts.Contacts`. The result will a) have `Mutual` flag b) will be easily comparable with previous one in append mode.

What do you think? Do you prefer `TL_contacts_contacts.Users`?

**Comment by patrizok** [comment]
1 - Actually there is no reason to reverse contacts, it doesn't matter, but I changed it to avoid misunderstanding.
2 - I run your code and I checked TL version downgraded to 126 in dump files.
3 - Let's leave it as it is for easy processing (**TL_contacts_contacts.Users**). If someone need **b** options, then it can be changed.

**Comment by 3bl3gamer** [comment]
Cool! Thank you again, will merge it now.

I will also merge some changes soon which will fix log colors on Windows. If it won't work, fell free to open another issue!

**Pull request merged**